File: MATLAB/CalcRollResonance.m
% Main script for running calculations. Note that RollResonance.m must
% be in the same folder as this script.

obj = RollResonance("CN_A.txt", "CL_Delta.txt", "CL_P.txt", "CMQ.txt");
obj.InitInputData("Dynamic_Roll.txt");
obj.InitializeVars();

eps = 0.00436;      % rad
gamma = 22.5;       % deg
static_AoA = 0.05;   % deg
cg_offset = 0.005;  % m
Toffset = 0.001;    % m
mu = 0;             % deg

fin_cant_angle = 0.35;

obj.SetRefQuantities(0.205, 0.03301, fin_cant_angle)
obj.AssumeQuantities(eps, gamma, Toffset, cg_offset, static_AoA, mu);
obj.FindNaturalFrequency();

obj.InitializeBC(100, -4.3138E-15, -0.003732135, 0, 2.144275746)
for i = 101:length(obj.Time)-1
    % Roll equation via Laplace Transform
    obj.CalcDynamicRoll(i, 100); 
    
    % Steady state equation
    obj.CalcSteadyStateRoll(i);
    
    if (i <= 2100)
       obj.CalcRoots(i);
       obj.CalcComplexPlane(i);
    end
    
    % Just to show calculations are running
    if(mod(i,100) ==0)
       fprintf("Iteration %d00, calculating...\n", i/100);
   end
end

x_limit = round(max(abs(obj.sideslip(101:2100))));
y_limit = round(max(abs(obj.AoA(101:2100))));

graph_limit = x_limit + 1;
if (y_limit > x_limit)
    graph_limit = y_limit + 1;
end

figure(1);
plot(obj.Time, obj.omega);
hold on;
plot(obj.Time, obj.P/360);
plot(obj.Time, obj.Pss/360);
plot(obj.Time, obj.PssCalc/360);
xlabel('Time (s)', 'fontsize', 12);
ylabel('Frequency (Hz)', 'fontsize', 12);
xlim([0 100]);
ylim([0 8]);
legend('Natural Frequency','Dynamic - Laplace Transform', ...
    'ASTOS Output', 'Steady State Equation');
grid on;
hold off;

figure(2);
plot(obj.POmegaRatio, obj.trimAoA/obj.staticAoA);
xlabel('P / \omega', 'fontsize', 12);
ylabel('\alpha_{TRIM} / \alpha_{STATIC}', 'fontsize', 12);
grid on
xlim([0 2]);

figure(3);
ax = gca;
% Calculated sideslip and AoA
plot(obj.sideslip(101:2100), obj.AoA(101:2100));
hold on;
ax.DataAspectRatio = [1 1 1];

% Astos Output
plot(obj.AstosSS(101:2100), obj.AstosAoA(101:2100));

% Trim Arm
plot(obj.trimArmReal(101:2100), obj.trimArmImag(101:2100));

plot([-graph_limit graph_limit], [0 0], 'k');
plot([0 0], [-graph_limit graph_limit], 'k');
grid on;
xlabel('Sideslip Angle (\circ)', 'fontsize', 12);
ylabel('Angle of Attack (\circ)', 'fontsize', 12);
legend('Tricyclic Theory', 'ASTOS Output', 'Trim Arm');
title("Static AoA: " + static_AoA + "\circ, CG Offset: " + obj.CGoffset*1000 + ...
    " mm, Thrust Offset: " + obj.Toffset*1000 + " mm", 'fontsize', 14);
x0=50;
y0=50;
width=800;
height=800;
set(gcf,'position',[x0,y0,width,height])
hold off;

figure(4);
plot(obj.POmegaRatio, obj.phase);
xlabel('P / \omega', 'fontsize', 12);
ylabel('Phase Angle (\circ)', 'fontsize', 12);
grid on
xlim([0 2]);

File: Python/CalcRollResonance.py
import numpy as np
import math
import cmath
import matplotlib.pyplot as plt
import RollResonance
from math import sin, radians
from scipy.interpolate import interp1d, interp2d

# ------------------------------------------------------------------------------------------------------------------
x = RollResonance.RollResonance("CN_A.txt", "CL_Delta.txt", "CL_P.txt", "CMQ.txt")
x.InitInputData("Dynamic_Roll.txt")

# Diameter, Reference area, Cant angle
fin_cant = 0.35
x.SetRefQuantities(0.205, 0.03306, fin_cant)
x.FindNaturalFreq()
x.InitializeVars()

static_AoA = 0.05
cg_offset = 0.005
Toffset = 0.001
title = "Static AoA: " + str(static_AoA) + "°, CG Offset: " + str(cg_offset) + "m, Thrust Offset: " + str(Toffset) + "m"

# eps(rad), gamma(deg), Toffset(m), CGoffset(m), staticAoA(deg), mu
x.AssumeQuantities(0.00436, 22.5, Toffset, cg_offset, static_AoA, 0)

x.InitializeBC(100, -4.3138E-15, -0.003732135, 0, 2.144275746)

for i in range(101, x.time.size-1):
    x.CalcDynamicRoll(i, 101)
    x.CalcSteadyRoll(i)
    if (i <= 2100):
        x.CalcRoots(i)
        x.CalcComplexPlane(i)

x.CalcClContributions()

# ------------------------------------------------------------------------------------------------------------------
# Figure plots
plt.figure()
plt.plot(x.PomegaRatio[101:2100], x.trimAoA[101:2100]/x.staticAoA)
plt.axis([0,2,0,30])
plt.ylabel(r'$\alpha_{TRIM}/\alpha_{STATIC}$', fontsize=12)
plt.xlabel(r'$P/\omega$', fontsize=12)
plt.grid()
plt.savefig("TrimAoA",bbox_inches = 'tight', pad_inches = 0.3)

plt.clf()
plt.plot(x.time, x.PssCalc/360,label='Steady-State Equation')
plt.plot(x.time, x.P/360,label='Dynamic - Laplace Transform')
plt.plot(x.time, x.omega, label='Natural Frequency')
plt.plot(x.time, x.Pss/360, label="Astos")
plt.ylabel('Frequency (Hz)', fontsize=12)
plt.xlabel('Time (s)', fontsize=12)
plt.axis([0,100,0,8])
plt.grid()
plt.legend()
plt.savefig("Pitch_Roll_Freq",bbox_inches = 'tight', pad_inches = 0.3)

plt.clf()
plt.plot(x.time, x.P/360,label=str(fin_cant)+'\N{DEGREE SIGN} Fin Cant')
plt.plot(x.time, x.omega, label='Natural Frequency')
plt.ylabel('Frequency (Hz)', fontsize=12)
plt.xlabel('Time (s)', fontsize=12)
plt.axis([0,100,0,12])
plt.grid()
plt.legend()
plt.savefig("Design_Chart_Roll_Freq",bbox_inches = 'tight', pad_inches = 0.3)

plt.clf()
plt.plot(x.PomegaRatio[101:2100], x.CLacg[101:2100],label='Effect due to ACG')
plt.plot(x.PomegaRatio[101:2100], x.CLss[101:2100], label='Effect due to Cl_p and Cl_delta')
plt.ylabel(r'$C_L$', fontsize=12)
plt.xlabel(r'$P/\omega$', fontsize=12)
plt.axis([0,1.2,-0.04,0.12])
plt.grid()
plt.legend()
plt.savefig("CL",bbox_inches = 'tight', pad_inches = 0.3)

plt.clf()
plt.plot(x.PomegaRatio, x.phase)
plt.grid()
plt.axis([0,2,-90,90])
plt.ylabel("Phase Angle (\N{DEGREE SIGN})", fontsize=12)
plt.xlabel(r'$P/\omega$', fontsize=12)
plt.savefig("Phase_Angle",bbox_inches = 'tight', pad_inches = 0.3)

plt.clf()
fig, ax = plt.subplots(figsize=(12, 10))
ax.axhline(y=0, color='k')
ax.axvline(x=0, color='k')
ax.set_aspect(1)
plt.plot(x.Sideslip[101:2100], x.AoA[101:2100], label="Tricyclic")
plt.plot(x.trimArmReal[101:2100], x.trimArmImag[101:2100], label="Trim Arm")
plt.plot(x.AstosSS[101:2100], x.AstosAoA[101:2100], label="ASTOS Output")
plt.ylabel("Angle of Attack (\N{DEGREE SIGN})", fontsize=12)
plt.xlabel("Sideslip Angle (\N{DEGREE SIGN})", fontsize=12)
plt.legend(fontsize=12)
plt.title(title, fontweight="bold", fontsize=14)
plt.grid()
plt.savefig("AoA_Sideslip",bbox_inches = 'tight', pad_inches = 0.3)
